examples/pennylane_run_doubly_stochastic.py

```diff
@@ -0,0 +1,312 @@
+
+plt.ylabel("Cost function value")
+plt.xlabel("Optimization steps")
+plt.legend()
+plt.show()
+
+##############################################################################
+# Using the trained parameters from each optimization strategy, we can
+# evaluate the analytic quantum device:
+
+print("Vanilla gradient descent min energy = ", qnode_GD(params_GD))
+print("Stochastic gradient descent (shots=100) min energy = ", qnode_GD(params_SGD100))
+print("Stochastic gradient descent (shots=1) min energy = ", qnode_GD(params_SGD1))
+
+##############################################################################
+# Amazingly, we see that even the ``shots=1`` optimization converged
+# to a reasonably close approximation of ground-state energy!
+
+##############################################################################
+# Doubly stochastic gradient descent for VQE
+# ------------------------------------------
+#
+# As noted in `Sweke et al. (2019) <https://arxiv.org/abs/1910.01155>`__,
+# variational quantum algorithms often include terms consisting of linear combinations
+# of expectation values. This is true of the parameter-shift rule (where the
+# gradient of each parameter is determined by shifting the parameter by macroscopic
+# amounts and taking the difference), as well as VQE, where the Hamiltonian
+# is usually decomposed into a sum of Pauli expectation values.
+#
+# Consider the Hamiltonian from the previous section. As this Hamiltonian is an
+# Hermitian observable, we can always express it as a sum of Pauli matrices using
+# the relation
+#
+# .. math::
+#
+#     H = \sum_{i,j=I,x,y,z} a_{i,j} (\sigma_i\otimes \sigma_j),
+#         ~~a_{i,j} = \frac{1}{4}\text{tr}[(\sigma_i\otimes \sigma_j )H].
+#
+# Applying this, we can see that
+#
+# .. math::
+#
+#     H = 4  + 2I\otimes X + 4I \otimes Z - X\otimes X + 5 Y\otimes Y + 2Z\otimes X.
+#
+# To perform "doubly stochastic" gradient descent, we simply apply the stochastic
+# gradient descent approach from above, but in addition also uniformly sample
+# a subset of the terms for the Hamiltonian expectation at each optimization step.
+# This inserts another element of stochasticity into the system---all the while
+# convergence continues to be guaranteed!
+#
+# Let's create a QNode that randomly samples three terms from the above
+# Hamiltonian as the observable to be measured.
+
+I = np.identity(2)
+X = np.array([[0, 1], [1, 0]])
+Y = np.array([[0, -1j], [1j, 0]])
+Z = np.array([[1, 0], [0, -1]])
+
+terms = np.array(
+    [2 * np.kron(I, X), 4 * np.kron(I, Z), -np.kron(X, X), 5 * np.kron(Y, Y), 2 * np.kron(Z, X)]
+)
+
+
+@qml.qnode(dev_SGD)
+def sampled_terms(params, A=None):
+    StronglyEntanglingLayers(*params, wires=[0, 1])
+    return expval(qml.Hermitian(A, wires=[0, 1]))
+
+
+def loss(params):
+    n = np.random.choice(np.arange(5), size=3, replace=False)
+    A = np.sum(terms[n], axis=0)
+    return 4 + (5 / 3) * sampled_terms(params, A=A)
+

```

Comment:
Maybe we could have something along the lines of 
```python
# After the tensor PR is merged this would be replaced with PL tensor operators
# qml.Tensor(qml.Identity, qml.PauliZ, wires=[0, 1]) etc.
terms = [qml.Hermitian(2 * np.kron(I, X), wires=[0,1]), ...]

@qml.qnode(dev_SGD)
def circuit(params, n):
    idx = np.random.choice(np.arange(5), size=n, replace=False)
    StronglyEntanglingLayers(*params, wires=[0, 1])
    
    return [expval(terms[i]) for i in idx]
```

In this way we would save the loss function and can still evaluate for `n = 5`.

Reply:
I've moved the sampling of the terms into the QNode, but unfortunately the terms list cannot be composed of quantum observables (as this must always be inside a QNode).

Reply:
Ah I see! Just another reason to revise the `do_queue` handling...